File: frontend/src/lib/api/validation.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { apiClient } from "@/lib/api/client";

export interface EvaluationMetric {
  metric_key: string;
  dimension_key: string | null;
  numeric_value: number | null;
  json_value: Record<string, unknown> | null;
  sample_count: number;
  calculation_version: string;
}

export interface ReleaseGateDecision {
  release_profile: string;
  status: string;
  zero_tolerance_failures: string[];
  metric_failures: string[];
}

export interface EvaluationRun {
  id: string;
  evaluation_dataset_id: string;
  status: string;
  items_evaluated: number;
  items_failed: number;
  created_at: string;
  completed_at: string | null;
  metrics: EvaluationMetric[];
  release_gate: ReleaseGateDecision | null;
}

export interface EvaluationTypeSummary {
  evaluation_type: string;
  latest_run: EvaluationRun | null;
}

export interface ValidationSummary {
  classification: EvaluationTypeSummary;
  retrieval: EvaluationTypeSummary;
  theme: EvaluationTypeSummary;
  grounding: EvaluationTypeSummary;
}

export function useValidationSummary() {
  return useQuery({
    queryKey: ["validation-summary"],
    queryFn: () => apiClient.get<ValidationSummary>("/api/v1/validation/summary"),
  });
}

export interface ReviewQueueItem {
  object_type: string;
  object_id: string;
  excerpt: string;
  dimension_key: string;
  dimension_display_name: string;
  label_key: string;
  label_display_name: string;
  confidence: number;
  previous_snapshot: Record<string, unknown>;
  created_at: string;
}

export interface ReviewQueueList {
  total_matching: number;
  limit: number;
  offset: number;
  items: ReviewQueueItem[];
}

/** Phase A of the review queue (design.md §32.4) — only `analysis_label`
 * is implemented so far; see `mellow-strolling-thimble.md` for the
 * remaining-phase roadmap (feedback_relevance, feedback_analysis, theme,
 * theme_membership, insight, insight_evidence, answer_finding,
 * answer_citation). */
export function useReviewQueue(objectType: string = "analysis_label") {
  return useQuery({
    queryKey: ["review-queue", objectType],
    queryFn: () =>
      apiClient.get<ReviewQueueList>(`/api/v1/validation/reviews?object_type=${objectType}`),
  });
}

export interface SubmitReviewRequest {
  object_type: string;
  object_id: string;
  decision: "accepted" | "edited" | "rejected" | "needs_second_review";
  previous_snapshot: Record<string, unknown>;
  edited_snapshot?: Record<string, unknown>;
  reason_code?: string;
  notes?: string;
}

export interface ReviewDecision {
  id: string;
  object_type: string;
  object_id: string;
  decision: string;
  reason_code: string | null;
  notes: string | null;
  created_at: string;
}

export function useSubmitReview() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (body: SubmitReviewRequest) =>
      apiClient.post<ReviewDecision>("/api/v1/validation/reviews", body),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ["review-queue"] }),
  });
}

File: backend/src/instamart_engine/api/schemas/validation.py
"""Request/response DTOs for the validation workspace API. architecture.md §20.7."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class EvaluationMetricResponse(BaseModel):
    metric_key: str
    dimension_key: str | None
    numeric_value: float | None
    json_value: dict[str, Any] | None
    sample_count: int
    calculation_version: str


class ReleaseGateDecisionResponse(BaseModel):
    release_profile: str
    status: str
    zero_tolerance_failures: list[str]
    metric_failures: list[str]


class EvaluationRunResponse(BaseModel):
    id: UUID
    evaluation_dataset_id: UUID
    status: str
    items_evaluated: int
    items_failed: int
    created_at: datetime
    completed_at: datetime | None
    metrics: list[EvaluationMetricResponse]
    release_gate: ReleaseGateDecisionResponse | None


class EvaluationTypeSummaryResponse(BaseModel):
    evaluation_type: str
    latest_run: EvaluationRunResponse | None


class ValidationSummaryResponse(BaseModel):
    classification: EvaluationTypeSummaryResponse
    retrieval: EvaluationTypeSummaryResponse
    theme: EvaluationTypeSummaryResponse
    grounding: EvaluationTypeSummaryResponse


class SubmitReviewRequest(BaseModel):
    object_type: str
    object_id: UUID
    decision: str
    previous_snapshot: dict[str, Any]
    edited_snapshot: dict[str, Any] | None = None
    reviewer_actor_id: UUID | None = None
    reason_code: str | None = None
    notes: str | None = Field(default=None, max_length=4000)


class ReviewDecisionResponse(BaseModel):
    id: UUID
    object_type: str
    object_id: UUID
    decision: str
    reason_code: str | None
    notes: str | None
    created_at: datetime


class ReviewQueueItemResponse(BaseModel):
    object_type: str
    object_id: UUID
    excerpt: str
    dimension_key: str
    dimension_display_name: str
    label_key: str
    label_display_name: str
    confidence: float
    previous_snapshot: dict[str, Any]
    created_at: datetime


class ReviewQueueListResponse(BaseModel):
    total_matching: int
    limit: int
    offset: int
    items: list[ReviewQueueItemResponse]
